pilot/pkg/model/jwks_resolver_test.go

```diff
@@ -418,6 +351,56 @@ func TestJwtPubKeyRefreshWithNetworkError(t *testing.T) {
 	verifyKeyLastRefreshedTime(t, r, ms, false /* wantChanged */)
 }
 
+func TestJwtRefreshIntervalOnFailureAndResetToDefault(t *testing.T) {
+	r := newJwksResolver(
+		JwtPubKeyEvictionDuration,
+		100*time.Millisecond, /*RefreshInterval*/
+		2*time.Millisecond,   /*RefreshIntervalOnFailure*/
+		testRetryInterval,
+	)
+	defer r.Close()
+
+	ms := startMockServer(t)
+	defer ms.Stop()
+
+	// Configures the mock server to return error after the first request.
+	ms.ReturnErrorAfterFirstNumHits = 2
+	ms.ReturnSuccessAfterFirstNumHits = 20
+
+	mockCertURL := ms.URL + "/oauth2/v3/certs"
+	_, err := r.GetPublicKey("", mockCertURL)
+	if err != nil {
+		t.Fatalf("GetPublicKey(\"\", %+v) fails: expected no error, got (%v)", mockCertURL, err)
+	}
+
+	// note that calling refresh has the side effect that it does hit the mockserver and hence we will already do the 2nd hit.
+	refreshInverval := r.refresh()
+	if refreshInverval != r.refreshDefaultInterval {
+		t.Errorf("first test on refresh interval failed, expected: %s, got: %s", r.refreshDefaultInterval, refreshInverval)
+	}
+
+	// sleep a bit more than the default refresh interval so we are expected to operate the refresh in failure mode.
+	time.Sleep(r.refreshDefaultInterval + 1*time.Millisecond)
+	refreshInverval = r.refresh()
```

Comment:
I didn't want to access `r.refreshInterval` directly because it would introduce a data race (refresher in separate goroutine is writing and this test in another goroutine is reading). That said I also think the whole test is a bit tricky as we can't directly observe the refresh interval and also cannot directly control it.

@yangminzhu Do you think we could change `JwksResolver` adding a RW mutex which would make this test much easier. With adding the RW mutex we could then also consider swapping out the `sync.Map` and use a normal map guarded by the RW mutex. I think the runtime cost for the RW mutex in this case should be negligible and the more ergonomic use of a "normal" map should out pay the runtime cost.

```golang
// JwksResolver is resolver for jwksURI and jwt public key.
type JwksResolver struct {
	mu sync.RWMutex

	// Callback function to invoke when detecting jwt public key change.
	PushFunc func()

	// cache for JWT public key.
	keyEntries map[jwtKey]jwtPubKeyEntry
```

Reply:
I've rewritten the unit test for the refresh behavior with using a lock on `r.refreshInterval` and calling refresh directly instead of passively observe it.